### PR TITLE
sparse/highdim

### PR DESCRIPTION
**Pull request by zitongzhan**
Solve https://github.com/pypose/pypose/issues/236 
This PR is branched from https://github.com/pypose/pypose/pull/232. Thus #232 needs to be merged first, before this PR can be merged.

### CODE REVIEW

**Review by huyaoyu — COMMENTED** [review]
Algorithm is good.

Just need to revise things like naming, comments, and doc string

File: pypose/sparse/sparse_block_tensor.py [hunk 1]
```diff
@@ -163,7 +181,7 @@ def repeated_value_as_hybrid_value(coo: torch.Tensor,
         raise ValueError(f'Unknown mode: {mode}')
     val = val[..., None, None].tile(block_shape)
 
-    s_dim: List[int] = list(coo.shape[-2:])
+    s_dim: List[int] = list(coo.shape)
 
     return torch.sparse_coo_tensor(
```
[the remark this answers is not in the document]
Comment (zitongzhan) [reply pypose/sparse/sparse_block_tensor.py:186]: Yes. It is no longer used.

File: pypose/sparse/sparse_block_tensor.py [hunk 2]
```diff
@@ -225,44 +244,45 @@ def coo_2_hybrid(coo, proxy, dense_proxy_thres: int=DENSE_SAFE_PROXY_THRES):
...
-    b_dim_t = torch.tensor(b_dim, dtype=torch.int64, device=coo.device).unsqueeze(-1)
-    in_block_indices = coo.indices() % b_dim_t
+    shape_b_t = torch.tensor(shape_b, dtype=torch.int64, device=coo.device).unsqueeze(-1)
+    inblock_indices = coo.indices() % shape_b_t
     # block indices.
-    block_indices = coo.indices() // b_dim_t
-    numel_proxy = torch.numel(proxy)
-    if numel_proxy < dense_proxy_thres:  # check the *dense* shape
+    indices_b = coo.indices() // shape_b_t
+    numel_p = torch.numel(proxy)
```
Comment (huyaoyu) [inline pypose/sparse/sparse_block_tensor.py:259]: Can put an additional comment here saying `torch.numel()` will return the dense number of element, not the sparse number.

**Review by huyaoyu — APPROVED** [review]
Looks good. Only some minor comments. Let's then try to create more unit tests to make sure we cover high-dimensional cases.

File: pypose/sparse/sparse_block_tensor.py [hunk 3]
```diff
@@ -109,46 +125,48 @@ def make_coo_indices_and_dims_from_hybrid(hybrid):
     hybrid = hybrid.coalesce()
 
     # The block dimension.
-    assert len(hybrid.shape) % 2 == 0
-    b_dim = len(hybrid.shape) // 2
-    b_shape = hybrid.shape[b_dim:]
-    p_shape = hybrid.shape[:b_dim]
-    assert hybrid.dim() >= 4, f'hybrid.shape = {hybrid.shape}. '
-    n_block_elem = prod(b_shape)
+    assert len(hybrid.shape) % 2 == 0, 'The hybrid tensor must have even number of dims, ' \
```
[the remark this answers is not in the document]
Comment (zitongzhan) [reply pypose/sparse/sparse_block_tensor.py]: yes, replaced them with `hybrid.dim()`
